File: packages/sdk-typescript/src/core/sse.ts
export interface SseEvent {
  readonly id?: string;
  readonly event?: string;
  readonly data: string;
}

interface Accumulator {
  id?: string;
  event?: string;
  data: string[];
}

/**
 * Parse a byte stream of Server-Sent Events into frames, following the W3C event
 * stream format: `field: value` lines, `data` lines joined with newlines, a blank
 * line dispatching the accumulated event, and `:` comment lines ignored. Frames
 * with no data are not emitted (e.g. lone comments or keep-alive colons).
 *
 * Line terminators `\n` and `\r\n` are handled, including a `\r\n` split across
 * chunk boundaries. Lone `\r` terminators are not used by SSE servers and are not
 * treated as line breaks.
 */
export async function* parseSse(stream: ReadableStream<Uint8Array>): AsyncGenerator<SseEvent> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  let frame: Accumulator = { data: [] };

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let newline: number;
      // biome-ignore lint/suspicious/noAssignInExpressions: idiomatic incremental line scan
      while ((newline = buffer.indexOf('\n')) !== -1) {
        let line = buffer.slice(0, newline);
        buffer = buffer.slice(newline + 1);
        if (line.endsWith('\r')) line = line.slice(0, -1);

        if (line === '') {
          const event = flush(frame);
          if (event) yield event;
          frame = { data: [] };
          continue;
        }
        if (line.startsWith(':')) continue;

        const colon = line.indexOf(':');
        const field = colon === -1 ? line : line.slice(0, colon);
        let value = colon === -1 ? '' : line.slice(colon + 1);
        if (value.startsWith(' ')) value = value.slice(1);

        if (field === 'event') frame.event = value;
        else if (field === 'data') frame.data.push(value);
        else if (field === 'id') frame.id = value;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

function flush(frame: Accumulator): SseEvent | null {
  if (frame.data.length === 0) return null;
  const event: { id?: string; event?: string; data: string } = { data: frame.data.join('\n') };
  if (frame.id !== undefined) event.id = frame.id;
  if (frame.event !== undefined) event.event = frame.event;
  return event;
}

File: packages/sdk-python/src/bimpeai/_sse.py
from __future__ import annotations

import codecs
from collections.abc import AsyncIterator, Iterator
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SseEvent:
    data: str
    id: str | None = None
    event: str | None = None


class _Decoder:
    """Incrementally turn decoded SSE text into events (W3C event stream format)."""

    def __init__(self) -> None:
        self._buffer = ""
        self._id: str | None = None
        self._event: str | None = None
        self._data: list[str] = []

    def feed(self, chunk: str) -> list[SseEvent]:
        self._buffer += chunk
        out: list[SseEvent] = []
        while True:
            newline = self._buffer.find("\n")
            if newline == -1:
                break
            line = self._buffer[:newline]
            self._buffer = self._buffer[newline + 1 :]
            if line.endswith("\r"):
                line = line[:-1]
            event = self._line(line)
            if event is not None:
                out.append(event)
        return out

    def _line(self, line: str) -> SseEvent | None:
        if line == "":
            event = self._flush()
            self._id = None
            self._event = None
            self._data = []
            return event
        if line.startswith(":"):
            return None
        colon = line.find(":")
        if colon == -1:
            field, value = line, ""
        else:
            field = line[:colon]
            value = line[colon + 1 :]
            if value.startswith(" "):
                value = value[1:]
        if field == "event":
            self._event = value
        elif field == "data":
            self._data.append(value)
        elif field == "id":
            self._id = value
        return None

    def _flush(self) -> SseEvent | None:
        if not self._data:
            return None
        return SseEvent(data="\n".join(self._data), id=self._id, event=self._event)


def parse_sse(chunks: Iterator[bytes]) -> Iterator[SseEvent]:
    decoder = _Decoder()
    text = codecs.getincrementaldecoder("utf-8")()
    for chunk in chunks:
        yield from decoder.feed(text.decode(chunk))


async def aparse_sse(chunks: AsyncIterator[bytes]) -> AsyncIterator[SseEvent]:
    decoder = _Decoder()
    text = codecs.getincrementaldecoder("utf-8")()
    async for chunk in chunks:
        for event in decoder.feed(text.decode(chunk)):
            yield event
